### PR TITLE
Fixes #7474 - Handles all redirects for Web UI File CRUD

### LINKED ISSUE #7474: Editing file and selecting to create a branch & make pull request doesn't go to pull request page
When you edit a file in the web interface and choose to create a branch and start a PR when you commit changes, it just goes back to viewing the file you edited, rather than a create PR page.

### PR DESCRIPTION
**Pull request by richmahn**
This fixes Issue #7474 and #7492. (backported to v1.9 in #7489 ) 

This PR enhances the commit form to better handle branching and making a patch and going to the pull request creation page once done with a file create/update/delete/upload (it wasn't going to the pull request create page if you selected to make a new branch).

This also fixes how deleting a file in a directory was taking you back to the root, when you should stay in the directory, or go up to the first one that has files.

### CODE REVIEW

File: routers/repo/editor.go [hunk 1]
```diff
@@ -636,3 +659,37 @@ func RemoveUploadFileFromServer(ctx *context.Context, form auth.RemoveUploadFile
 	log.Trace("Upload file removed: %s", form.File)
 	ctx.Status(204)
 }
+
+// GetUniquePatchBranchName Gets a unique branch name for a new patch branch
+// It will be in the form of <lowername>-patch-<num> where <num> is the first branch of this format
+// that doesn't already exist
+func GetUniquePatchBranchName(ctx *context.Context) string {
+	prefix := ctx.User.LowerName + "-patch-"
+	for i := 1; i <= 1000; i++ {
```
Comment (kolaente) [inline routers/repo/editor.go:674]: What if there are more than 1000 patch branches already? I doubt that would be likely, but what if?

File: routers/repo/editor.go [hunk 2]
```diff
@@ -636,3 +659,37 @@ func RemoveUploadFileFromServer(ctx *context.Context, form auth.RemoveUploadFile
...
+		branchName := fmt.Sprintf("%s%d", prefix, i)
+		if _, err := ctx.Repo.Repository.GetBranch(branchName); err != nil {
+			if git.IsErrBranchNotExist(err) {
+				return branchName
+			}
+			return ""
```
Comment (kolaente) [inline routers/repo/editor.go:681]: Could you log the error in that case?

File: routers/repo/editor.go [hunk 3]
```diff
@@ -636,3 +659,37 @@ func RemoveUploadFileFromServer(ctx *context.Context, form auth.RemoveUploadFile
...
+// GetClosestParentWithFiles Recursively gets the path of parent in a tree that has files (used when file in a tree is
+// deleted). Returns "" for the root if no parents other than the root have files
+func GetClosestParentWithFiles(treePath string, commit *git.Commit) string {
+	if len(treePath) == 0 || treePath == "." {
+		return ""
+	}
+	// see if the tree has entries
+	if tree, err := commit.SubTree(treePath); err != nil {
+		// failed to get tree, going up a dir
+		return GetClosestParentWithFiles(filepath.Dir(treePath), commit)
```
Comment (kolaente) [inline routers/repo/editor.go:697]: Please correct me if I'm wrong, but where do you go up a dir here? To me, it looks like this is doing exactly the same thing as in L692.

File: routers/repo/editor.go (hunk 1 above)
Comment (richmahn) [reply routers/repo/editor.go:674]: Then there will be just an empty field for the branch name and the user will have to put that in themselves, just like it is now.

File: routers/repo/editor.go (hunk 2 above)
Comment (richmahn) [reply routers/repo/editor.go:681]: Sure.

File: routers/repo/editor.go (hunk 3 above)
Comment (richmahn) [reply routers/repo/editor.go:697]: filepath.Dir(treePath) goes up a dir.

Comment (richmahn) [reply routers/repo/editor.go:697]: Here if we for some reason don't get a SubTree for this treePath we go up one more and try again. L692 is if we get a SubTree but there are no entries in it.

File: routers/repo/editor.go (hunk 2 above)
Comment (richmahn) [reply routers/repo/editor.go:681]: Done.

File: routers/repo/editor.go (hunk 3 above)
Comment (richmahn) [reply routers/repo/editor.go:697]: I added that to the function description.

File: routers/repo/editor.go (hunk 1 above)
Comment (richmahn) [reply routers/repo/editor.go:674]: I added that to the function description

File: routers/repo/editor.go (hunk 3 above)
Comment (kolaente) [reply routers/repo/editor.go:697]: :+1:

**Review by kolaente — APPROVED** [review]

File: routers/repo/editor_test.go [hunk 4]
```diff
@@ -5,6 +5,8 @@
 package repo
 
 import (
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/test"
```
Comment (mrsdizzie) [inline routers/repo/editor_test.go:9]: These should go below with "code.gitea.io/gitea/models" per normal styling preferences

**Review by mrsdizzie — CHANGES_REQUESTED** [review]
Small nit, this LGTM otherwise

**Review by mrsdizzie — APPROVED** [review]

**Pull request merged**